src/crowdin_client.cpp

```diff
@@ -143,139 +158,331 @@ dispatch::future<void> CrowdinClient::Authenticate()
-    data.add_value("import_eq_suggestions", "0");
-    data.add_file("files[" + str::to_utf8(file) + "]", "upload.po", file_content);
-
-    return m_api->post(url, data).then([](json){});
+    wxLogTrace("poedit.crowdin", "UploadFile(project_id=%ld, lang=%s, file_id=%ld, file_extension=%s)", project_id, lang.LanguageTag().c_str(), file_id, file_extension.c_str());
+    return m_api->post(
+            "storages",
+            octet_stream_data(file_content),
+            { { "Crowdin-API-FileName", "crowdin." + file_extension } }
+        )
+        .then([this, project_id, file_id, lang] (json r) {
+            wxLogTrace("poedit.crowdin", "File uploaded to temporary storage: %s", r.dump().c_str());
+            const auto storageId = r["data"]["id"].get<int>();
+            return m_api->post(
+                "projects/" + std::to_string(project_id) + "/translations/" + lang.LanguageTag(),
+                json_data({
+                    { "storageId", storageId },
+                    { "fileId", file_id }
+                }))
+                .then([](json r) {
+                    wxLogTrace("poedit.crowdin", "File uploaded: %s", r.dump().c_str());
+                });
+        });
 }
 
 
 bool CrowdinClient::IsSignedIn() const
 {
     std::string token;
-    return keytar::GetPassword("Crowdin", "", &token);
+    return keytar::GetPassword("Crowdin", "", &token)
+        && token.substr(0, 2) == "2:";
 }
 
 
 void CrowdinClient::SignInIfAuthorized()
 {
     std::string token;
-    if (keytar::GetPassword("Crowdin", "", &token))
+    if (keytar::GetPassword("Crowdin", "", &token)
+        && token.length() > 2)
+    {
+        token = token.substr(2);
         SetToken(token);
+    }
+    wxLogTrace("poedit.crowdin", "Token: %s", token.c_str());
+}
+
+
+static std::string base64_decode_json_part(const std::string &in) {
+
+    std::string out;
+    std::vector<int> t(256, -1);
+    {
+        static const char* b = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+        for (int i = 0; i < 64; i ++)
+            t[b[i]] = i;
+    }
+
+    int val = 0,
+        valb = -8;
+
+    for (uint8_t c : in)
+    {
+        if (t[c] == -1)
+            break;
+        val = (val << 6) + t[c];
+        valb += 6;
+        if (valb >= 0)
+        {
+            out.push_back(char(val >> valb & 0xFF));
+            valb -= 8;
+        }
+    }
+
+    return out;
 }
 
 
 void CrowdinClient::SetToken(const std::string& token)
 {
-    m_api->set_authorization(!token.empty() ? "Bearer " + token : "");
+    wxLogTrace("poedit.crowdin", "Authorization: %s", token.c_str());
+
+    if(token.empty())
+        return;
+
+    std::string domain;
+    try
+    {
+        domain = json::parse(
+            base64_decode_json_part(std::string(
+                wxString(token).AfterFirst('.').BeforeFirst('.').utf8_str()
+        ))).at("domain").get<std::string>();
+        domain += ".";
+    }
+    catch(...) {}
```

Comment:
Don't you think this situation - unexpectedly bad token - warrants at least trace logging, given how logging-heavy all the rest of the changes are? An explanatory comment along the lines of "assume token is non-enterprise, fall through with empty domain" would help in understanding that it is intentional and *correct* that the error is ignored.

Reply:
Addressed cc589293a710f0f0ecfaed767de94f3fe829edd0 . Sorry that have not implemented it initially, just hurried up to get features themself working and testable.